lib/Language/Haskell/Stylish/Step/Imports.hs

```diff
@@ -160,6 +201,63 @@ formatImports maxCols options moduleStats rawGroup =
   forM_ group \imp -> printQualified options padNames stats imp >> newline
 
 
+--------------------------------------------------------------------------------
+groupAndFormat
+  :: Maybe Int
+  -> Options
+  -> ImportStats
+  -> [NonEmpty (GHC.LImportDecl GHC.GhcPs)]
+  -> Editor.Edits
+groupAndFormat maxCols options moduleStats groups = edits
+  where edits = Editor.changeLines block (const regroupedLines)
+        regroupedLines =
+          intercalate [""] $
+          map (formatImports maxCols options moduleStats) grouped
+
+        grouped = groupByPatterns (groupPatterns options) imports
+        imports = concatMap toList groups
+
+        block = Block
+          (GHC.srcSpanStartLine . src $ head imports)
+          (GHC.srcSpanEndLine   . src $ last imports)
+        src = fromMaybe (error "regroupImports: missing location") .
+          GHC.srcSpanToRealSrcSpan . GHC.getLocA
+
+groupByPatterns
+  :: [Pattern]
+  -> [GHC.LImportDecl GHC.GhcPs]
+  -> [NonEmpty (GHC.LImportDecl GHC.GhcPs)]
+groupByPatterns patterns allImports = go patterns allImports []
+  where
+    go [] [] groups            = groups
+    go [] imports groups       = groups <> [NonEmpty.fromList imports]
+    go (p : ps) imports groups =
+      let
+        (groups', rest) = extract p imports
+      in
+        go ps rest (groups <> groups')
+
+    extract p imports =
+      let
+        (matched, rest) =
+          partition (matches p) imports
+        subgroups =
+          groupBy ((==) `on` capture p) $ sortOn (capture p) matched
```

Comment:
It looks like we're going to run `p` on the same text 3 times doing this. But regex matching is fast so probably not worth worrying about.

Reply:
Yep, exactly my reasoning. I started out trying to write a version that only ran the pattern once, but it ended up making the logic quite a bit more complex, and I decided that it wasn't worth it. Matching with regex-tdfa should be fast and I expect module names would break GHC before getting long enough to cause performance problems here.

Reply:
Perhaps worth a short comment pointing that out.